### DISCUSSION TITLE
Can't Set up Google Social Auth

### DISCUSSION THREAD

**Discussion by osseonews** [nhost/nhost#2734] [Q&A]
Your instructions for Google Social Auth here, https://docs.nhost.io/guides/auth/social/sign-in-google, don't work. The reason is because you say that "Under Authorized JavaScript origins, add your project’s redirect URL in the following format". But, your URL is a nhost url. If you add to the Javascript origins in the Credentials, then the Oauth Consent Screen automatically adds the nhost.run domain to the list of Authorized Domains in the Oauth Screen step 1. 

However, you need to verify ownership of each domain listed there to get approved. Of course, we cannot verify ownership of the nhost.run domain, as we don't own it!  So how do we actually set up the social auth? It seems impossible and I have no idea how anyone has set this up, as you can't verify owners of the nhost.run domain. Can we leave out the "Under Authorized JavaScript origins, add your project’s redirect UR"? Honestly, when we set up another Social Auth with another service provider, we didn't use the Javascript origin, just the Authorized redirect URIs section and it works fine. So why is Nhost asking to add the Javascript Origin? You can't actually create a production app, if you do this, due to the verification requirements per above.

Please help. Thanks.

**Comment by osseonews** [comment]
To follow up on the above, when use the useProvider hook, it creates a google link as follows: SUBDOMAIN.auth.REGION.nhost.run/v1/signin/provider/google , but we cannot possibly verify ownership of nhost.run domain, so you cannot use this as the original URL for the request. The origin URL has to come from our website which we can verify ownership for. Honestly, I am baffled as to how anyone has set this up with Nhost. I've done this many times with other Auth providers, so I have some idea as to how to do it, but this has me stumped

**Reply by osseonews** [reply]
And here is the screenshot from Google when we try to login with Gmail. The app is asking to login to continue with Nhost.run! But, this is not our website. So the set up is totally wrong and it's impossible to use. The user is NOT signing up with Nhost.run. They are signing up with our website. How do you fix this?


![nhost-run-google](https://github.com/nhost/nhost/assets/26583659/41656f09-7663-4956-b360-def4915583c1)

**Reply by osseonews** [reply]
About to give up here and go back to Clerk as this set up makes zero sense. You can't have the callback/redirect URL for Google Auth be subdomain.auth.region.nhost.run/v1/signin/provider/google. this is nonsensical. Then Google redirects to Nhost to login (per the screenshot above), but the user is NOT logging into nhost.run, they are logging into the domain we own. So the redirect has to be on our domain. But, if we redirect to our domain, which we tried, how do we authenticate them on nhost after Google redirects them? We can't because every single authentication has some redirect thru nhost.

**Comment by osseonews** [comment]
So my only solution here is to use this package: https://github.com/MomenSherif/react-oauth which is extremely simple to set up and runs on your own domain. The problem is that once the user is logged in with Google and you have decoded the credentialResponse, how do you log them into Nhost? You can't use passwordless login, as then the user needs to verify their email again. So which hook do you use to log in a user, if they are already logged in with an external system? Anonymous also doesn't work, as we need to assign the email and username to the login. And how do you prevent requirement to verify email?

**Comment by dbarrosop** [comment]
Google only requests to verify ownership under certain circumstances. Circumstances only they know, of course. In any case, you can use [custom domains](https://docs.nhost.io/platform/custom-domains) if you want to use your own domain instead of an nhost.run one.

**Reply by osseonews** [reply]
This is not true. You can't possibly use a redirect from nhost to login into google on a production domain because the redirect is coming from nhost, so Google will show the login as coming from nhost, per the screen shot above, not our domain. And you cannot get approved for production if the domain is not verified. Anyway, even if we could possibly verify nhost.run, it's again not a proper solution, because we are not going to have our users go to nhost.run to login, which is what Google will show. Custom domain is not an option for us.

**Reply by dbarrosop** [reply]
Yes, you can, we have lots of users doing it. I understand you may not want it for a production application but unless google has decided you need to verify the domain, which isn't a common thing, there is no technical or security issue with it.

Why is a custom domain not a solution? Isn't that what you want? If you don't want to use nhost.run not your own domain, what is it you want to use instead?

**Reply by osseonews** [reply]
OK, I see now how custom domains are done. I guess this is the only way to possibly do this as you can't redirect thru nhost.run for auth in a production scenario. Another way is to use react-oauth, but then you can't log in the user, because if the email is already taken you can't deanonmize the user.

**Reply by osseonews** [reply]
"verify the domain, which isn't a common thing." This is primary condition of going to production, so I have no idea why you think this is uncommon. If the domain is not verified, you can't use it. you think Google will allow any domain to be used for logging in to Oauth. This would be a massive security breach.

Anyway, even if you could possibly verify nhost.run, you can't possibly send your users there to login. It seems like a sort of phishing attack. 

So yeah, I'm guessing the only way to run this in production is a custom domain.

**Reply by osseonews** [reply]
Also, you got to add this to the documentation for Social Auth. Something like: "Add a Custom Domain for production, so the origin URL matches the verified domain." I didn't even know a custom domain was possible for auth and I literally wasted 6 hours already trying to figure this out with Google.

**Reply by osseonews** [reply]
Trying to add a custom domain but verify and save is disabled? How do you actually add a custom domain? how do you get the form to save?

**Reply by dbarrosop** [reply]
do you have a connected repo? AFAIK that's the only case where it might be disabled to avoid inconsistencies. If that's the case you can use the nhost.toml. If you have a connected repo but aren't using the config file send an email to support and we can enable it for you.

PS: there should be a banner at the top with the reason

**Reply by osseonews** [reply]
We don't have a connected repo. Sent an email to support. Don't see any banner.

**Comment by dbarrosop** [comment]
Also, we'd appreciate if you could
be a bit more mindful with your language. Thanks.

**Reply by osseonews** [reply]
"mindful" what is wrong with my language exactly? Google auth simply doesn't work because you cannot possibly have a redirect from another unauthorized domain. Other people have posted about it. I'm trying to figure out a solution. No clue what is wrong with what I wrote.
